File: python/src/oppex_sdk/_internal/dispatcher.py
"""Internal bounded asynchronous dispatcher. Not part of the supported SDK API."""

from __future__ import absolute_import

import logging
import threading
import time

from oppex_sdk._compat import monotonic, queue_module

LOGGER = logging.getLogger(__name__)

WORKER_COUNT = 2
QUEUE_CAPACITY = 5000
SHUTDOWN_TIMEOUT_SECONDS = 10.0
DROP_LOG_INTERVAL_SECONDS = 60.0

_POLL_INTERVAL_SECONDS = 0.1
_FORCED_JOIN_SECONDS = 1.0
_CLOSED_MESSAGE = "IncidentClient is closed"


class RateLimitedDropLogger(object):
    """Reports dropped incidents at most once per interval.

    A saturated queue drops continuously, so per-drop logging would replace one
    overload with another.
    """

    def __init__(self, logger=None, clock=None, interval_seconds=DROP_LOG_INTERVAL_SECONDS):
        self._logger = LOGGER if logger is None else logger
        self._clock = time.time if clock is None else clock
        self._interval_seconds = interval_seconds
        self._lock = threading.Lock()
        self._interval_started_at = self._clock()
        self._dropped_in_interval = 0

    def record_drop(self):
        with self._lock:
            self._dropped_in_interval += 1
            now = self._clock()
            if now - self._interval_started_at < self._interval_seconds:
                return
            self._interval_started_at = now
            count = self._dropped_in_interval
            self._dropped_in_interval = 0
        if count > 0:
            self._logger.warning("Dropped %d incidents in the last %g seconds.",
                                 count, self._interval_seconds)


class _TrackedTask(object):
    """Keeps the queue depth accurate whether a task runs or is dropped."""

    def __init__(self, delegate, metrics, drop_logger):
        self._delegate = delegate
        self._metrics = metrics
        self._drop_logger = drop_logger
        self._lock = threading.Lock()
        self._started_or_dropped = False

    def run(self):
        if not self._claim():
            return
        self._metrics.decrement_queued()
        self._delegate()

    def drop(self):
        if not self._claim():
            return
        self._metrics.decrement_queued()
        self._metrics.increment_dropped()
        self._drop_logger.record_drop()

    def _claim(self):
        with self._lock:
            if self._started_or_dropped:
                return False
            self._started_or_dropped = True
            return True


class AsyncDispatcher(object):
    """Runs best-effort deliveries on a fixed pool of daemon worker threads.

    The queue is bounded. When it is full the oldest queued incident is dropped
    so the newest one still gets a chance, and submission never blocks the
    calling application.
    """

    def __init__(self, metrics, shutdown_signal=None, worker_count=WORKER_COUNT,
                 queue_capacity=QUEUE_CAPACITY, drop_logger=None,
                 shutdown_timeout_seconds=SHUTDOWN_TIMEOUT_SECONDS):
        self._metrics = metrics
        self._shutdown_signal = threading.Event() if shutdown_signal is None else shutdown_signal
        self._drop_logger = RateLimitedDropLogger() if drop_logger is None else drop_logger
        self._shutdown_timeout_seconds = shutdown_timeout_seconds
        self._queue = queue_module.Queue(maxsize=queue_capacity)
        self._stopping = threading.Event()
        self._close_lock = threading.Lock()
        self._closed = False
        self._submit_lock = threading.Lock()
        self._workers = [self._start_worker(index + 1) for index in range(worker_count)]

    def submit(self, task):
        """Queues ``task``, dropping the oldest queued task when the queue is full."""
        if task is None:
            raise ValueError("task must not be None")
        if self._stopping.is_set():
            raise RuntimeError(_CLOSED_MESSAGE)

        tracked = _TrackedTask(task, self._metrics, self._drop_logger)
        self._metrics.increment_queued()
        with self._submit_lock:
            if self._offer(tracked):
                return
            oldest = self._poll()
            if oldest is not None:
                oldest.drop()
            if self._stopping.is_set() or not self._offer(tracked):
                tracked.drop()

    def close(self):
        """Drains queued work for a bounded period, then drops whatever is left."""
        with self._close_lock:
            if self._closed:
                return
            self._closed = True

        self._stopping.set()
        deadline = monotonic() + self._shutdown_timeout_seconds
        for worker in self._workers:
            worker.join(max(0.0, deadline - monotonic()))

        if any(worker.is_alive() for worker in self._workers):
            # The graceful window expired: stop in-flight retry backoff rather
            # than delaying application shutdown any further.
            self._shutdown_signal.set()
        # Nothing may stay queued once a close returns.
        self._drain_pending()
        for worker in self._workers:
            worker.join(_FORCED_JOIN_SECONDS)

    def _start_worker(self, sequence):
        worker = threading.Thread(target=self._work, name="oppex-incident-worker-%d" % sequence)
        worker.daemon = True
        worker.start()
        return worker

    def _work(self):
        while True:
            try:
                task = self._queue.get(True, _POLL_INTERVAL_SECONDS)
            except queue_module.Empty:
                if self._stopping.is_set():
                    return
                continue
            try:
                task.run()
            except Exception:
                # A worker must outlive any single failed delivery.
                LOGGER.debug("Oppex incident worker task failed", exc_info=True)

    def _drain_pending(self):
        while True:
            task = self._poll()
            if task is None:
                return
            task.drop()

    def _offer(self, tracked):
        try:
            self._queue.put_nowait(tracked)
            return True
        except queue_module.Full:
            return False

    def _poll(self):
        try:
            return self._queue.get_nowait()
        except queue_module.Empty:
            return None

File: rust/src/internal/dispatcher.rs
use std::collections::VecDeque;
use std::fmt;
use std::sync::{Arc, Condvar, Mutex, MutexGuard};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant};

use crate::internal::drop_logger::{DROP_LOG_INTERVAL, RateLimitedDropLogger};

pub(crate) const QUEUE_CAPACITY: usize = 5000;
pub(crate) const WORKER_COUNT: usize = 2;
pub(crate) const CLOSE_DRAIN_TIMEOUT: Duration = Duration::from_secs(10);

pub(crate) type Task = Box<dyn FnOnce() + Send + 'static>;

/// Everything the workers and the closing thread coordinate on lives behind one
/// mutex. Keeping the lifecycle flags inside it rather than in separate atomics
/// is what makes the "check the flag, then wait" sequence race-free: a worker
/// cannot miss the wake-up that close sends between those two steps.
struct State {
    tasks: VecDeque<Task>,
    /// Whether new work is still admitted. Cleared first thing by `close`.
    accepting: bool,
    /// Whether workers should keep draining. Cleared once `close` gives up.
    draining: bool,
    /// Tasks popped from the queue but not yet finished.
    active: usize,
}

/// Delivers asynchronous incidents on a fixed number of worker threads, queueing
/// the rest up to `capacity` and dropping the oldest entry once full.
///
/// Dropping the oldest keeps the newest incident, which is the one most likely to
/// still matter, and submission never blocks the application.
///
/// `Debug` is written by hand rather than derived: a queued task is a boxed
/// closure, which cannot be `Debug`, and every public type in this crate should
/// still be printable.
pub(crate) struct AsyncDispatcher {
    shared: Arc<Shared>,
    workers: Mutex<Vec<JoinHandle<()>>>,
}

struct Shared {
    capacity: usize,
    state: Mutex<State>,
    changed: Condvar,
    drop_logger: RateLimitedDropLogger,
}

impl fmt::Debug for AsyncDispatcher {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let state = self.shared.lock();
        formatter
            .debug_struct("AsyncDispatcher")
            .field("capacity", &self.shared.capacity)
            .field("queued", &state.tasks.len())
            .field("active", &state.active)
            .field("accepting", &state.accepting)
            .finish_non_exhaustive()
    }
}

impl AsyncDispatcher {
    pub(crate) fn new(worker_count: usize, capacity: usize) -> Self {
        let shared = Arc::new(Shared {
            capacity,
            state: Mutex::new(State {
                tasks: VecDeque::new(),
                accepting: true,
                draining: true,
                active: 0,
            }),
            changed: Condvar::new(),
            drop_logger: RateLimitedDropLogger::new(DROP_LOG_INTERVAL),
        });

        let workers = (0..worker_count)
            .map(|index| {
                let shared = Arc::clone(&shared);
                thread::Builder::new()
                    .name(format!("oppex-async-{index}"))
                    .spawn(move || shared.work())
                    .expect("spawning an Oppex worker thread")
            })
            .collect();

        Self {
            shared,
            workers: Mutex::new(workers),
        }
    }

    /// Queues a task, evicting the oldest queued task when the queue is full.
    /// Reports whether the task was accepted; only a closed dispatcher refuses.
    pub(crate) fn submit(&self, task: Task) -> bool {
        let dropped = {
            let mut state = self.shared.lock();
            if !state.accepting {
                return false;
            }
            let dropped = if state.tasks.len() >= self.shared.capacity {
                state.tasks.pop_front();
                self.shared.drop_logger.record_drop()
            } else {
                None
            };
            state.tasks.push_back(task);
            dropped
        };
        self.shared.changed.notify_one();

        if let Some(dropped) = dropped {
            log::warn!("oppex: dropped {dropped} incidents in the last minute");
        }
        true
    }

    /// Stops admitting work and drains what is already queued and in flight for
    /// up to `timeout`, then abandons the rest. Idempotent, and safe to call
    /// from any thread.
    pub(crate) fn close(&self, timeout: Duration) {
        {
            let mut state = self.shared.lock();
            if !state.accepting {
                return;
            }
            state.accepting = false;
        }

        let abandoned = self.shared.drain_until(Instant::now() + timeout);
        self.shared.changed.notify_all();

        // The workers are deliberately not joined. A thread cannot be
        // interrupted in Rust, so joining one that is blocked in an HTTP attempt
        // would push `close` past its own timeout by up to the attempt timeout.
        // Java's `close()` makes the same trade: it waits for the drain, then
        // calls `shutdownNow()` and returns without waiting for whatever that
        // failed to interrupt. An idle worker exits as soon as it sees
        // `draining == false`; a busy one exits when its current attempt ends,
        // which the attempt timeout already bounds.
        drop(
            self.workers
                .lock()
                .unwrap_or_else(std::sync::PoisonError::into_inner),
        );

        // Reported once and directly, rather than through the rate-limited
        // overflow counter: a shutdown loss and an overload loss are different
        // events, and a rate-limited counter's last batch can go unreported.
        if abandoned > 0 {
            log::warn!("oppex: force-dropped {abandoned} pending incidents during close");
        }
    }
}

impl Shared {
    fn lock(&self) -> MutexGuard<'_, State> {
        // A poisoned lock means a task panicked. The queue itself is still
        // consistent, so recovering keeps one bad task from wedging delivery
        // and shutdown for the whole process.
        self.state
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner)
    }

    fn work(&self) {
        loop {
            let task = {
                let mut state = self.lock();
                loop {
                    if let Some(task) = state.tasks.pop_front() {
                        state.active += 1;
                        break Some(task);
                    }
                    if !state.draining {
                        return;
                    }
                    state = self
                        .changed
                        .wait(state)
                        .unwrap_or_else(std::sync::PoisonError::into_inner);
                }
            };

            if let Some(task) = task {
                task();
                self.lock().active -= 1;
                self.changed.notify_all();
            }
        }
    }

    /// Waits until nothing is queued or in flight, or the deadline passes. Then
    /// stops the workers and returns how many tasks were abandoned.
    fn drain_until(&self, deadline: Instant) -> usize {
        let mut state = self.lock();
        while !state.tasks.is_empty() || state.active > 0 {
            let Some(remaining) = deadline.checked_duration_since(Instant::now()) else {
                break;
            };
            let (next, _) = self
                .changed
                .wait_timeout(state, remaining)
                .unwrap_or_else(std::sync::PoisonError::into_inner);
            state = next;
        }
        state.draining = false;
        let abandoned = state.tasks.len();
        state.tasks.clear();
        abandoned
    }
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::mpsc::{channel, sync_channel};
    use std::sync::{Arc, Mutex};
    use std::time::{Duration, Instant};

    use super::AsyncDispatcher;

    #[test]
    fn runs_submitted_work() {
        let dispatcher = AsyncDispatcher::new(2, 10);
        let (sender, receiver) = channel();

        for index in 0..5 {
            let sender = sender.clone();
            assert!(dispatcher.submit(Box::new(move || sender.send(index).unwrap())));
        }
        drop(sender);
        dispatcher.close(Duration::from_secs(5));

        let mut delivered: Vec<i32> = receiver.iter().collect();
        delivered.sort_unstable();
        assert_eq!(vec![0, 1, 2, 3, 4], delivered);
    }

    #[test]
    fn drops_the_oldest_queued_task_when_full() {
        // One worker, held on a gate, so everything else has to queue.
        let dispatcher = AsyncDispatcher::new(1, 2);
        let (release, held) = sync_channel::<()>(0);
        let (started, worker_started) = sync_channel::<()>(0);
        dispatcher.submit(Box::new(move || {
            started.send(()).unwrap();
            held.recv().unwrap();
        }));
        worker_started.recv().unwrap();

        let ran = Arc::new(Mutex::new(Vec::new()));
        for name in ["oldest", "middle", "newest"] {
            let ran = Arc::clone(&ran);
            dispatcher.submit(Box::new(move || ran.lock().unwrap().push(name)));
        }

        release.send(()).unwrap();
        dispatcher.close(Duration::from_secs(5));

        let ran = ran.lock().unwrap();
        assert_eq!(
            2,
            ran.len(),
            "the queue holds two tasks, so one must be dropped: {ran:?}"
        );
        assert!(
            !ran.contains(&"oldest"),
            "the oldest queued task must be the one dropped: {ran:?}"
        );
    }

    #[test]
    fn refuses_work_after_close_and_close_is_idempotent() {
        let dispatcher = AsyncDispatcher::new(1, 4);
        dispatcher.close(Duration::from_secs(1));
        dispatcher.close(Duration::from_secs(1));

        assert!(!dispatcher.submit(Box::new(|| unreachable!("a closed dispatcher must refuse work"))));
    }

    #[test]
    fn close_gives_up_on_work_that_outlasts_the_drain_timeout() {
        let dispatcher = AsyncDispatcher::new(1, 10);
        let (release, held) = sync_channel::<()>(0);
        let (started, worker_started) = sync_channel::<()>(0);
        dispatcher.submit(Box::new(move || {
            started.send(()).unwrap();
            held.recv().unwrap();
        }));
        worker_started.recv().unwrap();

        let ran = Arc::new(AtomicUsize::new(0));
        let counter = Arc::clone(&ran);
        dispatcher.submit(Box::new(move || {
            counter.fetch_add(1, Ordering::Release);
        }));

        let started_closing = Instant::now();
        dispatcher.close(Duration::from_millis(100));
        let elapsed = started_closing.elapsed();

        assert!(
            elapsed < Duration::from_secs(5),
            "close() waited {elapsed:?}, well past its timeout"
        );
        assert_eq!(0, ran.load(Ordering::Acquire), "an abandoned task must not run");
        drop(release);
    }
}
